app/scripts/lib/indexed-db-backend.ts

```diff
@@ -0,0 +1,185 @@
+  if (hashString !== checksum) {
+    throw new Error(`Checksum mismatch for key ${key}`);
+  }
+};
+
+export class IndexedDBBackend implements StorageBackend {
+  private storeName: string;
+
+  private dbVersion: number;
+
+  constructor(storeName: string, dbVersion: number) {
+    this.storeName = storeName;
+    this.dbVersion = dbVersion;
+  }
+
+  private async _getObjectStore(
+    mode: IDBTransactionMode,
+  ): Promise<IDBObjectStore> {
+    return new Promise((resolve, reject) => {
+      const request = indexedDB.open(this.storeName, this.dbVersion);
+
+      request.onerror = (event: Event) => {
+        reject(
+          new Error(
+            `Failed to open database ${this.storeName}: ${
+              (event.target as any)?.error
+            }`,
+          ),
+        );
+      };
+
+      request.onupgradeneeded = (event) => {
+        const db = (event.target as IDBOpenDBRequest).result;
+
+        if (!db.objectStoreNames.contains(this.storeName)) {
+          db.createObjectStore(this.storeName, {
+            keyPath: ['name', 'chainId'],
+          });
+        }
+      };
+
+      request.onsuccess = (event) => {
+        const db = (event.target as IDBOpenDBRequest).result;
+        const transaction = db.transaction([this.storeName], mode);
+        const objectStore = transaction.objectStore(this.storeName);
+        resolve(objectStore);
+      };
+    });
+  }
+
+  async read(key: StorageKey, checksum: string): Promise<ArrayBuffer> {
+    return new Promise<ArrayBuffer>((resolve, reject) => {
+      this._getObjectStore('readonly')
+        .then((objectStore) => {
+          const request = objectStore.get([key.name, key.chainId]);
+
+          request.onsuccess = async (event) => {
+            const data = (event.target as any)?.result?.data;
+            await validateChecksum(key, data, checksum).catch((error) => {
+              reject(error);
+            });
+            resolve(data);
+          };
+
+          request.onerror = (event) => {
+            reject(
+              new Error(`Error reading data: ${(event.target as any)?.error}`),
+            );
+          };
+        })
+        .catch((error) => {
+          reject(error);
+        });
+    });
+  }
+
+  async write(
+    key: StorageKey,
+    data: ArrayBuffer,
+    checksum: string,
+  ): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+      validateChecksum(key, data, checksum)
+        .then(() => {
+          this._getObjectStore('readwrite')
+            .then((objectStore) => {
+              const request = objectStore.put({ ...key, data });
+
+              request.onsuccess = () => {
+                resolve();
+              };
+
+              request.onerror = (event) => {
+                reject(
+                  new Error(
+                    `Error writing data: ${(event.target as any)?.error}`,
+                  ),
+                );
+              };
+            })
+            .catch((error) => {
```

Comment:
Why are we nesting here? Should we not validate then get, then create the handlers? That way we only need a single catch for the entire chain?

Reply:
Withdrawal transfer and deposit

Reply:
This is now resolved.